src/FSharp.Core/local.fs

```diff
@@ -480,17 +480,57 @@ module internal List =
             else
                 filterToFreshConsTail cons f t
 
-    let rec filter predicate l =
-        match l with
-        | [] -> l
-        | h :: ([] as nil) -> if predicate h then l else nil
-        | h :: t ->
-            if predicate h then
-                let cons = freshConsNoTail h
-                filterToFreshConsTail cons predicate t
-                cons
-            else
-                filter predicate t
+    [<System.Runtime.CompilerServices.MethodImpl(System.Runtime.CompilerServices.MethodImplOptions.NoInlining)>]
+    let invalidStateOfEmptyList () = raise (System.Exception "logic error; list should not be empty")
+
+    /// optimized filter if filtered items are all at the end of the list then that list is returned
+    /// the function entryToFilterToFreshConsTail was the original version of filter, and is used
+    /// in the case of there being multiple segments in the data
+    let filter predicate originalList =
+        let rec entryToFilterToFreshConsTail lst =
```

Comment:
Don't use nested recursive functions in FSharp.Core - lift them out to the top level so we know exactly how they're compiled.